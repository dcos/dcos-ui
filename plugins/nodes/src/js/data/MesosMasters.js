import React from "react";

import { mesosMastersLeader } from "./MesosMastersLeader";
import { mesosMastersHealth } from "./MesosMastersHealth";

import LeaderGrid from "../components/LeaderGrid";
import NonLeadersGrid from "../components/NonLeadersGrid";

function mastersInitialState() {
  return {
    leader: {
      hostPort: undefined,
      hostIp: undefined,
      version: undefined,
      electedTime: undefined,
      startTime: undefined,
      region: undefined
    },
    masters: undefined
  };
}

function filterLeader(leader, masters) {
  if (masters === undefined) {
    return undefined;
  }

  return masters.filter(master => master.host_ip !== leader.hostIp);
}

export function combineMasterData(leaderDataSource, mesosMastersHealth) {
  const mesosLeader$ = leaderDataSource().startWith(
    mastersInitialState().leader
  );
  const mesosMasters$ = mesosMastersHealth().startWith(undefined);

  return function() {
    return mesosLeader$.combineLatest(mesosMasters$, (leader, masters) => ({
      leader,
      masters: filterLeader(leader, masters)
    }));
  };
}

// This is an attempt of the mediator pattern without componentFromStream
export function connectMasterComponent(initialState, stream) {
  class MesosMasters extends React.Component {
    constructor() {
      super(...arguments);

      this.stream = this.props.stream();
      this.state = this.props.initialState();
    }

    componentDidMount() {
      this.subscription = this.stream.subscribe(this.setState.bind(this));
    }

    componentWillUnmount() {
      this.subscription.unsubscribe();
    }

    render() {
      const { masters, leader } = this.state;

      return (
        <div>
          <LeaderGrid leader={leader} />
          <NonLeadersGrid masters={masters} />
        </div>
      );
    }
  }

  return () => <MesosMasters initialState={initialState} stream={stream} />;
}

const mastersWithHealth = combineMasterData(
  mesosMastersLeader,
  mesosMastersHealth
);

export default connectMasterComponent(mastersInitialState, mastersWithHealth);
