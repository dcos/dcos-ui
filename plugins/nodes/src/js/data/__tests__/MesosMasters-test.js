import React from "react";
import renderer from "react-test-renderer";
import { IntlProvider } from "react-intl";
import { Observable } from "rxjs/Observable";
import enUS from "#SRC/js/translations/en-US.json";

import { connectMasterComponent, combineMasterData } from "../MesosMasters";

const leader = {
  hostPort: "127.1.2.3:8080",
  version: "2.9.1",
  electedTime: 1532340694.04573,
  startTime: 1232340694.04573,
  region: "us-east-1"
};

const nonLeader = [
  {
    host_ip: "192.168.0.1",
    health: 0,
    role: "master",
    healthDescription: {
      classNames: "text-success",
      key: "HEALTHY",
      sortingValue: 3,
      title: "Healthy",
      value: 0
    }
  },
  {
    host_ip: "192.168.0.2",
    health: 1,
    role: "master",
    healthDescription: {
      classNames: "text-danger",
      key: "UNHEALTHY",
      sortingValue: 0,
      title: "Unhealthy",
      value: 1
    }
  },
  {
    host_ip: "192.168.0.3",
    health: 2,
    role: "master",
    healthDescription: {
      classNames: "text-warning",
      key: "WAR",
      sortingValue: 2,
      title: "Warning",
      value: 2
    }
  },
  {
    host_ip: "192.168.0.4",
    health: 4,
    role: "master",
    healthDescription: {
      classNames: "text-mute",
      key: "NA",
      sortingValue: 1,
      title: "N/A",
      value: 3
    }
  }
];

function mastersInitialState() {
  return {
    leader: {
      hostPort: undefined,
      hostIp: undefined,
      version: undefined,
      electedTime: undefined,
      startTime: undefined,
      region: undefined
    },
    masters: undefined
  };
}

const TestableMasterComponent = ({ component }) => {
  return (
    <IntlProvider locale="en" messages={enUS}>
      {component}
    </IntlProvider>
  );
};

describe("LeaderGrid", function() {
  beforeEach(function() {
    Date.now = jest.fn(() => 1542340694);
  });

  afterEach(function() {
    Date.now.mockRestore();
  });

  it("renders with running status", function() {
    const initialState = mastersInitialState;

    const leaderData = () => Observable.of(leader);
    const healthData = () => Observable.of(nonLeader);
    const combinedData = combineMasterData(leaderData, healthData);
    const MasterNodesTab = connectMasterComponent(initialState, combinedData);

    expect(
      renderer
        .create(<TestableMasterComponent component={<MasterNodesTab />} />)
        .toJSON()
    ).toMatchSnapshot();
  });
});
