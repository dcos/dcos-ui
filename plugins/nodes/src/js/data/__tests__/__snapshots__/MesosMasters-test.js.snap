// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeaderGrid renders with running status 1`] = `
<div>
  <div
    className="container"
  >
    <div
      className="configuration-map"
    >
      <h1
        className="detail-view-section-heading detail-view-section-heading-primary configuration-map-heading flush-top"
      >
        Leader
      </h1>
      <div
        className="configuration-map-section"
      >
        <div
          className="configuration-map-row table-row"
        >
          <div
            className="configuration-map-label"
          >
            IP and Port
          </div>
          <div
            className="configuration-map-value"
          >
            127.1.2.3:8080
          </div>
        </div>
        <div
          className="configuration-map-row table-row"
        >
          <div
            className="configuration-map-label"
          >
            Region
          </div>
          <div
            className="configuration-map-value"
          >
            us-east-1
          </div>
        </div>
        <div
          className="configuration-map-row table-row"
        >
          <div
            className="configuration-map-label"
          >
            <span>
              Version
            </span>
          </div>
          <div
            className="configuration-map-value"
          >
            2.9.1
          </div>
        </div>
        <div
          className="configuration-map-row table-row"
        >
          <div
            className="configuration-map-label"
          >
            <span>
              Started
            </span>
          </div>
          <div
            className="configuration-map-value"
          >
            in 39 years
          </div>
        </div>
        <div
          className="configuration-map-row table-row"
        >
          <div
            className="configuration-map-label"
          >
            <span>
              Elected
            </span>
          </div>
          <div
            className="configuration-map-value"
          >
            in 49 years
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="container"
  >
    <div
      className="configuration-map"
    >
      <h1
        className="detail-view-section-heading detail-view-section-heading-primary configuration-map-heading flush-top"
      >
        Non-Leaders
      </h1>
      <div
        className="configuration-map-section"
      >
        <div
          className="configuration-map-row table-row"
        >
          <div
            className="configuration-map-label"
          >
            192.168.0.1
          </div>
          <div
            className="configuration-map-value"
          >
            <span
              className="text-success"
            >
              Healthy
            </span>
          </div>
        </div>
        <div
          className="configuration-map-row table-row"
        >
          <div
            className="configuration-map-label"
          >
            192.168.0.2
          </div>
          <div
            className="configuration-map-value"
          >
            <span
              className="text-danger"
            >
              Unhealthy
            </span>
          </div>
        </div>
        <div
          className="configuration-map-row table-row"
        >
          <div
            className="configuration-map-label"
          >
            192.168.0.3
          </div>
          <div
            className="configuration-map-value"
          >
            <span
              className="text-warning"
            >
              Warning
            </span>
          </div>
        </div>
        <div
          className="configuration-map-row table-row"
        >
          <div
            className="configuration-map-label"
          >
            192.168.0.4
          </div>
          <div
            className="configuration-map-value"
          >
            <span
              className="text-mute"
            >
              N/A
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
